### PR TITLE
fix #279417: Add brackets after HBox with *Create system header*

### PR DESCRIPTION
**Pull request by tobik**
This PR add brackets to the system after a HBox with *Create system header*. Contrary to standard brackets, which are anchored in front of the system, this brackets are anchored to a measure.

### CODE REVIEW

File: libmscore/layout.cpp [hunk 1]
```diff
@@ -3444,15 +3445,21 @@ System* Score::collectSystem(LayoutContext& lc)
                   ww  += rest * m->ticks().ticks() * stretch;
                   m->stretchMeasure(ww);
                   m->layoutStaffLines();
+                  if (createBrackets)
+                  {
```
Comment (mattmcclinch) [inline libmscore/layout.cpp]: This brace should be at the end of the previous line.

File: libmscore/layout.cpp [hunk 2]
```diff
@@ -3444,15 +3445,21 @@ System* Score::collectSystem(LayoutContext& lc)
...
+                        system->addBrackets(toMeasure(mb));
+                        createBrackets = false;
                   }
+            }
```
Comment (mattmcclinch) [inline libmscore/layout.cpp]: These 2 braces should each be indented 6 more spaces.

File: libmscore/layout.cpp [hunk 3]
```diff
@@ -3444,15 +3445,21 @@ System* Score::collectSystem(LayoutContext& lc)
...
             else if (mb->isHBox()) {
                   mb->setPos(pos + QPointF(toHBox(mb)->topGap(), 0.0));
                   mb->layout();
+                  createBrackets = toHBox(mb)->createSystemHeader();
                   }
             else if (mb->isVBox())
                   mb->setPos(pos);
             pos.rx() += ww;
-            }
+      }
```
Comment (mattmcclinch) [inline libmscore/layout.cpp]: Please move this brace back where it was.

File: libmscore/system.cpp [hunk 4]
```diff
@@ -361,6 +361,108 @@ void System::layoutSystem(qreal xo1)
...
+
+      qreal xoff2 = 0.0;         // x offset for instrument name
+
+      int columns = 0;
+      for (int idx = 0; idx < nstaves; ++idx) {
+            int c = 0;
+            for (auto bi : score()->staff(idx)->brackets())
+                  c = qMax(c, bi->column() + 1);
+            columns = qMax(columns, c);
+      }
```
Comment (mattmcclinch) [inline libmscore/system.cpp]: This brace should be indented 6 more spaces.

File: libmscore/system.cpp [hunk 5]
```diff
@@ -361,6 +361,108 @@ void System::layoutSystem(qreal xo1)
...
+            return;
+
+      int nstaves = _staves.size();
+
+      //---------------------------------------------------
+      //  find x position of staves
+      //    create brackets
+      //---------------------------------------------------
+
+      qreal xoff2 = 0.0;         // x offset for instrument name
```
Comment (mattmcclinch) [inline libmscore/system.cpp]: This variable is unused, and should be removed.

**Comment by tobik** [comment]
Thanks for your review - wilco.

**Comment by GaspardPO** [comment]
There are a lot of huge conditions, like :
```cpp
if (bl[k]->track() == track && bl[k]->column() == i && bl[k]->bracketType() == bi->bracketType() && bl[k]->measure() == NULL) {
```

I find hard to understand the meaning of such condition. May be you could extract it to a variable or a method and give it a name.

something like:
``` cpp
bool shouldDoSomething = bl[k]->track() == track && bl[k]->column() == i && bl[k]->bracketType() == bi->bracketType() && bl[k]->measure() == NULL;

if (shouldDoSomething) {
```

File: libmscore/system.cpp [hunk 6]
```diff
@@ -361,6 +361,106 @@ void System::layoutSystem(qreal xo1)
...
+      int nstaves = _staves.size();
+
+      //---------------------------------------------------
+      //  find x position of staves
+      //    create brackets
+      //---------------------------------------------------
+
+      int columns = 0;
+      for (int idx = 0; idx < nstaves; ++idx) {
+            int c = 0;
```
Comment (GaspardPO) [inline libmscore/system.cpp]: what does "c" stands for?

**Comment by tobik** [comment]
I generally agree. In this particular case it is mainly copy & paste from LayoutSystem(). I did not want to change this code to be close to the original.

File: libmscore/system.cpp (hunk 6 above)
Comment (tobik) [reply libmscore/system.cpp]: See my comment to your previous message. 
Variable names in MuseScore are also a mystery to me

**Comment by mattmcclinch** [comment]
> I generally agree. In this particular case it is mainly copy & paste from LayoutSystem(). I did not want to change this code to be close to the original.

It is usually better to refactor reusable code into a function that can be reused, rather than copy & paste large chunks of code. I wonder if that can be done in this case.

**Comment by tobik** [comment]
I did some refactoring to avoid c&p

**Comment by mattmcclinch** [comment]
Thank you for heeding my earlier comments. I like what you have done. The only other thing I see right now is that you have used `NULL` in a few places. I have been told that we prefer `nullptr`. Oh, and I image you would like to squash your commits as well.

**Comment by tobik** [comment]
I tried to squash but got a git error: 

```
>git reset --soft HEAD-4
fatal: ambiguous argument 'HEAD-4': unknown revision or path not in the working tree.
```

**Comment by tobik** [comment]
Squashed

File: libmscore/system.cpp [hunk 7]
```diff
@@ -226,49 +220,8 @@ void System::layoutSystem(qreal xo1)
...
-                                    b->setTrack(track);
-                                    }
-                              add(b);
-//                              if (bi->selected() != b->selected()) {
-//                                    bi->selected() ? score()->select(b) : score()->deselect(b);
-//                                    }
-                              b->setStaffSpan(firstStaff, lastStaff);
-                              bracketWidth[i] = qMax(bracketWidth[i], b->width());
-                              }
+                        Bracket* b = createBracket(bi, i, staffIdx, bl, nullptr);
```
Comment (dmitrio95) [inline libmscore/system.cpp]: Maybe passing the first system's measure to `createBracket()` here can prevent creation of extra brackets I described above?

File: libmscore/bracket.h [hunk 8]
```diff
@@ -41,6 +41,7 @@ class Bracket final : public Element {
       // horizontal scaling factor for brace symbol. Cannot be equal to magY or depend on h
       // because layout needs width of brace before knowing height of system...
       qreal _magx;
+      Measure* _measure;
```
Comment (dmitrio95) [inline libmscore/bracket.h]: Could you please add an initialization of this variable (`Measure* _measure = nullptr;`) to make its default value defined?

**Comment by anatoly-os** [comment]
@tobik could you please address the review? Besides, LGTM :)

**Comment by tobik** [comment]
Oups, I managed to miss this review, sorry. 
I have a workshop this weekend but I will look ASAP next week.

**Pull request merged**